Example/Tests/GalleryCoordinatorTests.swift

```diff
@@ -0,0 +1,132 @@
+//
+//  GalleryCoordinatorTests.swift
+//  GiniVision_Tests
+//
+//  Created by Enrique del Pozo Gómez on 3/2/18.
+//  Copyright © 2018 Gini GmbH. All rights reserved.
+//
+
+import XCTest
+@testable import GiniVision
+
+final class GalleryCoordinatorTests: XCTestCase {
+    
+    let galleryManager = GalleryManagerMock()
+    let selectedImageDocuments: [GiniImageDocument] = [
+        GiniImageDocument(data: Data(count: 1), imageSource: .external),
+        GiniImageDocument(data: Data(count: 2), imageSource: .external),
+        GiniImageDocument(data: Data(count: 3), imageSource: .external)
+    ]
+    lazy var coordinator = GalleryCoordinator(galleryManager: self.galleryManager,
+                                              giniConfiguration: GiniConfiguration.sharedConfiguration)
+    
+    var dummyImagePicker: ImagePickerViewController {
+        return ImagePickerViewController(album: galleryManager.albums[0],
+                                         galleryManager: galleryManager,
+                                         giniConfiguration: GiniConfiguration.sharedConfiguration)
+    }
+    
+    var dummyAlbumPicker: AlbumsPickerViewController {
+        return AlbumsPickerViewController(galleryManager: galleryManager)
+    }
+    
+    override func setUp() {
+        super.setUp()
+        coordinator = GalleryCoordinator(galleryManager: self.galleryManager,
+                                         giniConfiguration: GiniConfiguration.sharedConfiguration)
+    }
+    
+    func testGalleryCoordinatorStart() {
+        coordinator.start()
+        
+        XCTAssertTrue(galleryManager.isCaching, "gallery manager should have started caching")
+        XCTAssertEqual(coordinator.galleryNavigator.viewControllers.count, 2,
+                       "there should be 2 view controller in the navigator when the gallery coordinator starts")
+    }
+    
+    func testCloseGallery() {
+        let delegate = GalleryCoordinatorDelegateMock()
+        coordinator.delegate = delegate
+        selectImage(at: 0, in: galleryManager.albums[2]) { _ in
+            _ = self.coordinator.cancelButton.target?.perform(self.coordinator.cancelButton.action)
+            
+            XCTAssertTrue(delegate.didCancelGallery,
+                          "gallery image picking should be cancel after tapping cancel button")
+            XCTAssertTrue(self.coordinator.selectedImageDocuments.isEmpty,
+                          "selected image documents collection should be cleared after cancelling")
+        }
+    }
+    
+    func testOpenImages() {
+        let delegate = GalleryCoordinatorDelegateMock()
+        coordinator.delegate = delegate
+        
+        selectImage(at: 0, in: galleryManager.albums[2]) { _ in
```

Comment:
Isn't there a way to have the select image method not work async? I would try to avoid that when possible

Reply:
Unfortunately the image is loaded asynchronously and I have to wait for it. Although I don't like async operations in unit tests, I think this way is at least clear.

Reply:
Ideally, you'd want to avoid triggering such "extra features" in a unit test. You don't need the image to load so the unit test shouldn't trigger it. But that's in an ideal world. If it's a lot of effort to not load the images, keep it as is.

Reply:
The only way is to set an empty value in the dictionary and assign it later once the image has been fetched. IMO this is quite raveled and could lead to big issues. In this case I prefer to keep it as it is.